tapgarden/re-org_watcher.go

```diff
@@ -0,0 +1,424 @@
+// carry out its duties.
+type ReOrgWatcherConfig struct {
+	// ChainBridge is the main interface for interacting with the chain
+	// backend.
+	ChainBridge ChainBridge
+
+	// SafeDepth is the number of confirmations we require before we
+	// consider a transaction to be safely buried in the chain.
+	SafeDepth uint32
+
+	// ErrChan is the main error channel the watcher will report back
+	// critical errors to the main server.
+	ErrChan chan<- error
+}
+
+// ReOrgWatcher is responsible for watching initially confirmed transactions
+// until they reach a safe confirmation depth. If a re-org happens, it will
+// update the proof and store it in the proof archive.
+type ReOrgWatcher struct {
+	startOnce sync.Once
+	stopOnce  sync.Once
+
+	cfg *ReOrgWatcherConfig
+
+	bestHeight uint32
+
+	incomingProofs chan *proofRegistration
+	incomingConfs  chan *chainntnfs.TxConfirmation
+
+	// pendingProofs is a list of all proofs that are currently being
+	// watched for re-orgs, keyed by their anchor transaction hash.
+	pendingProofs map[chainhash.Hash]*anchorTxNotification
+
+	// ContextGuard provides a wait group and main quit channel that can be
+	// used to create guarded contexts.
+	*fn.ContextGuard
+}
+
+// NewReOrgWatcher creates a new re-org watcher based on the passed config.
+func NewReOrgWatcher(cfg *ReOrgWatcherConfig) *ReOrgWatcher {
+	return &ReOrgWatcher{
+		cfg:            cfg,
+		incomingProofs: make(chan *proofRegistration),
+		incomingConfs:  make(chan *chainntnfs.TxConfirmation),
+		pendingProofs:  make(map[chainhash.Hash]*anchorTxNotification),
+		ContextGuard: &fn.ContextGuard{
+			DefaultTimeout: DefaultTimeout,
+			Quit:           make(chan struct{}),
+		},
+	}
+}
+
+// Start attempts to start a new re-org watcher.
+func (w *ReOrgWatcher) Start() error {
+	var startErr error
+	w.startOnce.Do(func() {
+		log.Info("Starting re-org watcher")
+
+		// Start the main event handler loop that will process new
+		// proofs and watch their anchor transactions until they reach
+		// a safe confirmation depth.
+		w.Wg.Add(1)
+		go w.watchTransactions()
+	})
+	return startErr
+}
+
+// Stop signals for a re-org watcher to gracefully exit.
+func (w *ReOrgWatcher) Stop() error {
+	var stopErr error
+	w.stopOnce.Do(func() {
+		log.Info("Stopping re-org watcher")
+
+		close(w.Quit)
+		w.Wg.Wait()
+	})
+
+	return stopErr
+}
+
+// watchTransactions processes new proofs given to the watcher and watches their
+// anchor transactions until they reach a safe confirmation depth.
+func (w *ReOrgWatcher) watchTransactions() {
+	defer w.Wg.Done()
+
+	runCtx, cancel := w.WithCtxQuitNoTimeout()
+	defer cancel()
+
+	newBlockChan, blockErr, err := w.cfg.ChainBridge.RegisterBlockEpochNtfn(
+		runCtx,
+	)
+	if err != nil {
+		w.reportErr(fmt.Errorf("unable to register for block "+
+			"epoch notifications: %w", err))
+		return
+	}
+
+	for {
+		select {
+		case newProofs := <-w.incomingProofs:
+			txHash := newProofs.proofs[0].AnchorTx.TxHash()
```

Comment:
Perhaps we can do a quick sanity check here that all the proofs have the same anchor tx hash?

Reply:
After that check, I think it might make sense to pass in `txHash` into `waitForConf` rather than select the first proof once more.

Reply:
The check could go in `WatchProofs()` to be as early as possible.

Related, AFAICT we only access that first proof / any proofs to look up the block height, block hash, or TXID for the group of proofs. We could store those 3 vars directly as members of `proofRegistration` instead of fetching them from a proof repeatedly.